Add static systemd unit files for workers

agrare opened the pull request:
Typically services have their systemd unit files installed by the packaging system, not dynamically at runtime.  Since these files don't change we should write them out statically at RPM install time rather than dynamically at startup.  Writing these dynamically was an artifact of systemd being an optional runtime mode in the past.

We still have to write the service `settings.conf` file because these settings are user configurable

TODO:
- [x] Add static unit files to plugins
- [x] Add unit file definitions to plugin generator

Related:
- [x] https://github.com/ManageIQ/manageiq-providers-amazon/pull/678
- [x] https://github.com/ManageIQ/manageiq-providers-ansible_tower/pull/252
- [x] https://github.com/ManageIQ/manageiq-providers-autosde/pull/55
- [x] https://github.com/ManageIQ/manageiq-providers-azure/pull/432
- [x] https://github.com/ManageIQ/manageiq-providers-azure_stack/pull/56
- [x] https://github.com/ManageIQ/manageiq-providers-dummy_provider/pull/36
- [x] https://github.com/ManageIQ/manageiq-providers-foreman/pull/84
- [x] https://github.com/ManageIQ/manageiq-providers-google/pull/178
- [x] https://github.com/ManageIQ/manageiq-providers-ibm_cloud/pull/131
- [x] https://github.com/ManageIQ/manageiq-providers-ibm_terraform/pull/58
- [x] https://github.com/ManageIQ/manageiq-providers-kubernetes/pull/421
- [x] https://github.com/ManageIQ/manageiq-providers-kubevirt/pull/181
- [x] https://github.com/ManageIQ/manageiq-providers-lenovo/pull/328
- [x] https://github.com/ManageIQ/manageiq-providers-nsxt/pull/34
- [x] https://github.com/ManageIQ/manageiq-providers-nuage/pull/240
- [x] https://github.com/ManageIQ/manageiq-providers-openshift/pull/203
- [x] https://github.com/ManageIQ/manageiq-providers-openstack/pull/685
- [x] https://github.com/ManageIQ/manageiq-providers-oracle_cloud/pull/21
- [x] https://github.com/ManageIQ/manageiq-providers-ovirt/pull/547
- [x] https://github.com/ManageIQ/manageiq-providers-redfish/pull/136
- [x] https://github.com/ManageIQ/manageiq-providers-scvmm/pull/180
- [x] https://github.com/ManageIQ/manageiq-providers-vmware/pull/692

Fryguy commented:
This will eliminate the need for root ultimately, right?

agrare commented:
This reduces the number of things that we require root for, but doesn't eliminate it.

We still need to be able to `systemctl enable service@guid` and write the `service@guid.d/override.conf` files which right now require root.

agrare commented:
@kbrock might have found a selinux policy to limit what `systemd api` commands we can run and possibly not require root for that
Still leaves writing out the service overrides, I need to investigate if there is a way to do that over the dbus api

agrare commented:
Thinking a rake task to build these would be helpful

Fryguy commented on app/models/miq_worker/systemd_common.rb:
```diff
@@ -202,5 +130,9 @@ def unit_environment_variables
     end
+
+    def systemd_unit_dir
+      Pathname.new("/usr/lib/systemd/system")
```
Why is this dir different than https://github.com/ManageIQ/manageiq/pull/20983/files#diff-67e01e3b1ea2575267fa04629e3e39801735cd8bac18b3608c1108e5fe3c602bR42 ?

agrare replied:
:+1: should be `/lib/systemd/system`

Fryguy commented on build/systemd_units/manageiq-ems_metrics_processor.target:
```diff
@@ -0,0 +1,2 @@
+[Unit]
+PartOf=miq.target
```
Where is this target file?  Also, can we use `manageiq.target`?

Fryguy replied:
I see it's been miq.target all along.  I'm fine with this PR being `miq.target` to avoid churn and/or upgrade issues, but wonder if we can change in the future.

agrare replied:
Now that we have a `manageiq-` prefix on these units I think `manageiq.target` makes more sense

agrare replied:
> to avoid churn and/or upgrade issues

We're already going to have to blow away the old files on upgrade to support the `manageiq-*` stuff in the rpm build so might as well change it

agrare commented:
Okay with the updated unit names the core workers are looking good:
```
[root@dhcp-8-99-105 vmdb]# systemctl status manageiq.slice
● manageiq.slice
   Loaded: loaded
   Active: active since Tue 2021-02-02 12:09:03 EST; 6min ago
    Tasks: 63
   Memory: 2.2G
   CGroup: /manageiq.slice
           ├─manageiq-event_handler.slice
           │ └─manageiq-event_handler@f7df775b-f18a-49b6-b384-229250a68819.service
           │   └─12204 MIQ: MiqEventHandler id: 52, queue: ems
           ├─manageiq-generic.slice
           │ ├─manageiq-generic@1449de5d-5543-4191-994f-6a7f391e7a48.service
           │ │ └─12217 MIQ: MiqGenericWorker id: 53, queue: generic
           │ └─manageiq-generic@a2838681-58fe-4ef7-822d-9287fc7b89a9.service
           │   └─12221 MIQ: MiqGenericWorker id: 54, queue: generic
           ├─manageiq-priority.slice
           │ ├─manageiq-priority@ac7dba22-29cb-4e35-afda-ad946a56105b.service
           │ │ └─12178 MIQ: MiqPriorityWorker id: 50, queue: generic
           │ └─manageiq-priority@d614b31a-e04e-4b40-a22a-b79bbc7ed3d1.service
           │   └─12191 MIQ: MiqPriorityWorker id: 51, queue: generic
           ├─manageiq-remote_console.slice
           │ └─manageiq-remote_console@c932c44a-513f-4450-9171-523a4b5478a5.service
           │   └─12281 puma 4.3.7 (tcp://127.0.0.1:5000) [MIQ: Web Server Worker]
           ├─manageiq-reporting.slice
           │ ├─manageiq-reporting@12e084f7-526d-496b-b91b-89ccc256ef53.service
           │ │ └─12231 MIQ: MiqReportingWorker id: 55, queue: reporting
           │ └─manageiq-reporting@1eac8af1-df92-4ffe-b94b-30814a1b424e.service
           │   └─12239 MIQ: MiqReportingWorker id: 56, queue: reporting
           ├─manageiq-schedule.slice
           │ └─manageiq-schedule@bdeeb0cc-1cd8-4dbd-b7bb-00feba905ab4.service
           │   └─12290 MIQ: MiqScheduleWorker id: 58
           ├─manageiq-ui.slice
           │ └─manageiq-ui@647247bc-f344-42ca-a8fa-7a56b16df305.service
           │   └─12307 puma 4.3.7 (tcp://127.0.0.1:3000) [MIQ: Web Server Worker]
           └─manageiq-web_service.slice
             └─manageiq-web_service@0b0efb74-07c3-4197-a71e-78de283728e9.service
               └─12323 puma 4.3.7 (tcp://127.0.0.1:4000) [MIQ: Web Server Worker]

Feb 02 12:09:03 dhcp-8-99-105.cloudforms.lab.eng.rdu2.redhat.com systemd[1]: Created slice manageiq.slice.
```

Still need to update the provider workers so that they'll be e.g. `manageiq-providers-amazon_cloud_refresh` instead of `manageiq-amazon_cloud_refresh`

agrare commented:
Okay got the majority of the provider workers going, need to do another pass to make sure I didn't miss any (amazon agent_coordinator_worker comes to mind)

```
[root@dhcp-8-99-105 vmdb]# systemctl status manageiq.slice
● manageiq.slice
   Loaded: loaded
   Active: active since Tue 2021-02-02 12:09:03 EST; 2h 5min ago
    Tasks: 86
   Memory: 3.2G
   CGroup: /manageiq.slice
           ├─manageiq-event_handler.slice
           │ └─manageiq-event_handler@5ef80903-472f-477c-ab3a-63dec6687c13.service
           │   └─15306 MIQ: MiqEventHandler id: 63, queue: ems
           ├─manageiq-generic.slice
           │ ├─manageiq-generic@1e06adf6-a7c7-43c6-bf5d-908aec48c167.service
           │ │ └─15322 MIQ: MiqGenericWorker id: 65, queue: generic
           │ └─manageiq-generic@90b3188a-6159-497c-8372-c4e6cca35728.service
           │   └─15319 MIQ: MiqGenericWorker id: 64, queue: generic
           ├─manageiq-priority.slice
           │ ├─manageiq-priority@728f4ed1-0f49-4e5e-b882-2418b55b2031.service
           │ │ └─15293 MIQ: MiqPriorityWorker id: 62, queue: generic
           │ └─manageiq-priority@a5aaaf45-a195-417b-ae9e-4c7fc9d2f3d8.service
           │   └─15280 MIQ: MiqPriorityWorker id: 61, queue: generic
           ├─manageiq-providers.slice
           │ ├─manageiq-providers-vmware_infra_manager_event_catcher.slice
           │ │ └─manageiq-providers-vmware_infra_manager_event_catcher@fbdf2daa-ba17-46e0-a570-7fadb74ddeec.service
           │ │   └─15864 MIQ: Vmware::InfraManager::EventCatcher id: 72, queue: ems_2
           │ ├─manageiq-providers-vmware_infra_manager_operations.slice
           │ │ └─manageiq-providers-vmware_infra_manager_operations@484f2c03-9ec0-4fc9-b573-3519e6d207a1.service
           │ │   └─15877 MIQ: Vmware::InfraManager::OperationsWorker id: 73, queue: ems_2
           │ └─manageiq-providers-vmware_infra_manager_refresh.slice
           │   └─manageiq-providers-vmware_infra_manager_refresh@be56e5bc-f834-486e-9555-f907fef0dc53.service
           │     └─15890 MIQ: Vmware::InfraManager::RefreshWorker id: 74, queue: ems_2
           ├─manageiq-remote_console.slice
           │ └─manageiq-remote_console@06ebc8f9-c7f0-4fe4-82a3-98c291ed4bf3.service
           │   └─15379 puma 4.3.7 (tcp://127.0.0.1:5000) [MIQ: Web Server Worker]
           ├─manageiq-reporting.slice
           │ ├─manageiq-reporting@1d1350ba-e2dc-4e4a-b47a-696a33f423c5.service
           │ │ └─15330 MIQ: MiqReportingWorker id: 66, queue: reporting
           │ └─manageiq-reporting@c4e5f138-0b44-4d22-acb3-dc41088e38a2.service
           │   └─15341 MIQ: MiqReportingWorker id: 67, queue: reporting
           ├─manageiq-schedule.slice
           │ └─manageiq-schedule@ac239c54-9bd5-4421-a883-ad219add8997.service
           │   └─15393 MIQ: MiqScheduleWorker id: 69
           ├─manageiq-ui.slice
           │ └─manageiq-ui@7db8a431-5696-4b58-9dc3-cf6cd723354f.service
           │   └─15408 puma 4.3.7 (tcp://127.0.0.1:3000) [MIQ: Web Server Worker]
           └─manageiq-web_service.slice
             └─manageiq-web_service@cb2f4ff7-6119-43eb-a5c4-ee3d51a7d524.service
               └─15430 puma 4.3.7 (tcp://127.0.0.1:4000) [MIQ: Web Server Worker]
```

Nice side benefit of `manageiq-providers*.slice` is we get a slice for just the provider workers.  Might investigate going one further and having a slice per provider-type like `/manageiq.slice/manageiq-providers.slice/manageiq-providers-vmware.slice`

kbrock commented on app/models/manageiq/providers/base_manager/event_catcher.rb:
```diff
@@ -19,4 +19,8 @@ def friendly_name
   end
+
+  def self.service_base_name
+    "manageiq-providers-#{minimal_class_name.underscore.tr("/", "_")}"
```
can we do a super class here?

agrare replied:
Yeah I considered that but some of these are queue workers and some are not so they have different parent classes.  Also considered a `ProviderWorkerMixin` but this would be the only thing we could share.  Maybe better to update `minimal_class_name` to include the `providers-` prefix?

agrare replied:
ProviderWorkerMixin came in handy here

kbrock commented:
Looks like our current `ExecStart` definition is a little different. I'm still working to make this easier to type

I put here rather than across all the repos
```
ExecStart=/bin/bash -c '. /etc/default/evm && exec ruby lib/workers/bin/run_single_worker.rb MiqEventHandler --heartbeat --guid=%i'
```

agrare commented:
Okay I updated this so that it doesn't co-depend on the rpm_build PR, this just adds the unit files

Fryguy commented:
@agrare So we can change the OP and remove the Depends/Dependent stuff (or rename to related or whatever)?

kbrock commented on build/systemd_units/manageiq-cockpit_ws@.service:
```diff
@@ -0,0 +1,11 @@
+[Service]
+WorkingDirectory=/var/www/miq/vmdb
+Environment=BUNDLER_GROUPS=manageiq_default,ui_dependencies
+ExecStart=/bin/bash -lc 'exec ruby lib/workers/bin/run_single_worker.rb MiqCockpitWsWorker --heartbeat --guid=%i'
```
Didn't get the evm environment completely into properties files, so can we go with this for now?
It is much more non-root friendly

```suggestion
ExecStart=/bin/bash -c '. /etc/default/evm && exec ruby lib/workers/bin/run_single_worker.rb MiqCockpitWsWorker --heartbeat --guid=%i'
```

agrare replied:
:+1: I'm good with that

The biggest thing holding this up is the rpm_build PR: https://github.com/ManageIQ/manageiq-rpm_build/pull/134

agrare replied:
Actually I'm going to hold off on this, the point of this change is to statically define what we currently write out dynamically.

We can make the change you're talking about independently of this change, either before or after it is merged.  If you want you can update the unit file template in `MiqWorker#systemd_common` and I'll re-generate these, or we can make that change to the service files after this is merged.

kbrock reviewed (commented):
Can we just make the one change and then merge this?
At a later date we will introduce the `User=miquser`.

If that ends up being too much work then we can keep this open until the commands are perfected

agrare commented:
Okay I've added systemd unit templates to the provider generator and added PRs for the new oracle_cloud provider and the dummy-provider

I left the content of these exactly as it would be on a current appliance (I don't want to change too many things at once re: https://github.com/ManageIQ/manageiq/pull/20983#discussion_r598988627)

This simply adds the files to the filesystem, we'll need https://github.com/ManageIQ/manageiq-rpm_build/pull/134 to copy them to the correct location on appliances then we can drop the code here which writes the files out
